plugins/api/api_main.ml

```diff
@@ -43,37 +103,95 @@ let main proj =
 
-type api = {lang : string; file : string}
+let api_of_paths paths =
+  List.concat (List.map ~f:Api_path.api_of_path paths)
+
+let show_all_apis paths =
+  let pp_api api =
+    Format.printf "   %s : %s@."
+      (Filename.basename api.file) api.file in
+  let print_lang apis = match apis with
+    | [] -> ()
+    | hd :: _ ->
+      Format.printf "@[<v>Language: %s@;" hd.lang;
+      List.iter ~f:pp_api apis;
+      Format.close_box ();
+      Format.print_newline () in
+  let apis = api_of_paths paths |>
+             List.sort ~cmp:(fun x y -> String.compare x.lang y.lang) in
+  List.iter ~f:print_lang (List.group ~break:(fun x y -> x.lang <> y.lang) apis);
+  Format.printf
+    "Total number of available API: %d\n" (List.length apis)
+
+let rec add_files apis paths = match paths with
+  | [] ->
+    eprintf "There are not any writable paths for api plugin\n";
+    exit 1
+  | path :: paths' ->
+    match apis with
+    | [] -> ()
+    | api :: apis' ->
+      match Api_path.cp path api with
+      | Ok () -> add_files apis' paths
+      | Error `No_source ->
+        eprintf "Api %s does not exist\n" api.file;
+        add_files apis' paths
+      | Error `Duplicated ->
+        eprintf "Api %s is already exists in api paths (%s), not added\n"
+          api.file path;
+        add_files apis' paths
+      | Error (`Fail reason) ->
+        eprintf "Failed to copy %s to %s because of error: %s\n"
+          api.file path reason;
+        add_files apis paths'
+
+let rem_files apis paths =
+  let rm api =
+    let found = List.fold ~init:[] ~f:(fun acc path ->
+        match Api_path.rm path api with
+        | Ok () -> path :: acc
+        | Error (`Fail reason) ->
+          eprintf "Api %s can't be removed from %s because of error %s\n"
+            api.file path reason;
+          path :: acc
+        | Error `No_source -> acc) paths in
+    if List.is_empty found then
+      eprintf "Api %s was not found in paths\n" api.file in
+  List.iter ~f:rm apis
 
-let add_files api =
-  let bundle = main_bundle () in
-  List.map api ~f:(fun api ->
-      Some ("api/" ^ api.lang ^ "/" ^ Filename.basename api.file),
-      Uri.of_string api.file) |>
-  Bundle.insert_files bundle
-
-let rem_files api =
-  main_bundle () |>
-  Bundle.update ~f:(fun file -> match String.split file ~on:'/' with
-      | ["api"; lang; file] ->
-        if List.exists api ~f:(fun x ->
-            x.lang = lang && x.file = Filename.basename file)
-        then `Drop
-        else `Copy
-      | "api" :: _ ->
-        warning "removing from the bundle a malformed entry: %s" file;
-        `Drop
-      | _ -> `Copy)
+let sanity_check paths =
+  let is_same api api' =
+    api.file <> api'.file && api.lang = api'.lang &&
+    Filename.basename api.file = Filename.basename api'.file in
+  let check_for_pair api apis =
+    match List.find ~f:(is_same api) apis with
+    | None -> ()
+    | Some api' ->
+      eprintf "sanity check failed for api plugin, apis with same \
```

Comment:
Ok, here we're requiring a user to fix the problem manually, so we need to be very verbose. 

```
Sanitization error: there are conflicting specifications for the same API(s):
<api> <path1> <path2>  ...
....

Each API should be specified only once. Please resolve the conflict manually, either by removing
the conflicting files or by renaming them. 
```